core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -2354,6 +2377,33 @@ boolean hasAliasUsedInGroupByWhichIsNotPresentInFinalProjection(Project rel) {
       return false;
     }
 
+    boolean hasFieldsUsedInFilterWhichIsNotUsedInFinalProjection(Project project) {
+      final SqlNodeList selectList = ((SqlSelect) node).getSelectList();
+      List<SqlNode> qualifyColumnList = new ArrayList<>();
+      ((SqlSelect) node).getQualify().accept(new SqlBasicVisitor<SqlNode>() {
+            @Override public SqlNode visit(SqlIdentifier id) {
+              qualifyColumnList.add(id);
+              return id;
+            }
+      });
+      if (selectList != null && qualifyColumnList != null) {
+        for (SqlNode grpNode : qualifyColumnList) {
+          if (grpNode instanceof SqlIdentifier) {
+            String grpCall = ((SqlIdentifier) grpNode).names.get(0);
+            for (SqlNode selectNode : selectList.getList()) {
```

Comment:
can we move this code to separate method to make it more readable.

private boolean isGrpCallNotUsedInFinalProjection(String grpCall, SqlNodeList selectList, Project project) {
    for (SqlNode selectNode : selectList.getList()) {
        if (selectNode instanceof SqlBasicCall) {
            if (grpCallIsAlias(grpCall, (SqlBasicCall) selectNode) && !grpCallPresentInFinalProjection(grpCall, project)) {
                return true;
            }
        }
    }
    return false;
}